src/control/server/storage/config.go

```diff
@@ -299,64 +299,120 @@ func (tcs TierConfigs) HaveBdevRoleMeta() bool {
+//     present, no bdev_roles: attributes are allowed on it.  This is the traditional
+//     PMem-based configuration, obviously not using MD-on-SSD.
 //
-//   - A role can only be assigned to one entire tier, i.e. tier 2 & 3 cannot both be assigned the
-//     Meta role. This doesn’t apply to the Data role which can be applied to multiple tiers e.g.
-//     in the case where > 3 bdev tiers exist.
-//   - All roles (WAL, Meta and Data) need to be assigned in bdev tiers if scm class is ram.
-//   - If the (first) scm tier is of class dcpm, then only one bdev tier with Data role is supported,
-//     no third tier (for now).
-func (tcs TierConfigs) validateBdevTierRoles() error {
-	sc := tcs.ScmConfigs()[0]
+//   - No storage tier with class: dcpm exists, exactly one storage tier with class:
+//     ram exists, and zero or one tier(s) with class: nvme exists. If an NVMe
+//     tier is present, no bdev_roles: attributes are specified for the NVMe tier.
+//     This is the traditional DRAM-based (ephemeral) configuration, and it shall not
+//     use MD-on-SSD (to be compatible/consistent with earlier levels of DAOS software).
+//
+//   - No storage tier with class: dcpm exists, exactly one storage tier with class:
+//     ram exists, and one, two or three tiers with class: nvme exist, with mandatory
+//     bdev_roles: attributes on each of the NVMe tiers. Each of the three roles
+//     (wal,meta,data) must be assigned to exactly one NVMe tier (no default
+//     assignments of roles to tiers by the control plane software; all roles shall be
+//     explicitly specified). This setup shall use the MD-on-SSD code path.  In this
+//     scenario allow the use of a single NVMe tier co-locating all three roles, three
+//     separate NVMe tiers with each tier dedicated to exactly one role, or two
+//     separate NVMe tiers where two of the three roles are co-located on one of the
+//     two NVMe tiers. In the latter case, all combinations to co-locate two of the
+//     roles shall be allowed, although not all those combinations may be technically
+//     desirable in production environments.
+//
+func (tcs TierConfigs) validateBdevRoles() error {
+	scmConfs := tcs.ScmConfigs()
+	if len(scmConfs) != 1 || scmConfs[0].Tier != 0 {
+		return errors.New("first storage tier is not scm")
+	}
+
+	sc := scmConfs[0]
 	bcs := tcs.BdevConfigs()
 
-	nrWALTiers := 1
-	nrMetaTiers := 1
-	if sc.Class == ClassDcpm {
-		nrWALTiers = 0
-		nrMetaTiers = 0
-		if len(bcs) > 1 {
-			return FaultBdevConfigMultiTiersWithDCPM
-		}
+	var wal, meta, data int
+	hasRoles := func() bool {
+		return wal > 0 || meta > 0 || data > 0
 	}
 
-	var foundWALTiers, foundMetaTiers, foundDataTiers int
-	for _, bc := range bcs {
-		bits := bc.Bdev.DeviceRoles.OptionBits
+	for i, bc := range bcs {
+		roles := bc.Bdev.DeviceRoles
+		if roles.IsEmpty() {
+			if hasRoles() {
+				return FaultBdevConfigRolesMissing
```

Comment:
From what I  understand, if the first bdev has no roles and the second one has some then the `FaultBdevConfigRolesMissing` will not be detected ?